crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,307 @@
+                .with_region(aws_region.to_string())
+                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                    credentials_provider: DefaultCredentialsChain::builder().build().await,
+                }))
+                .build()?;
+
+            Arc::new(store)
+        } else {
+            debug!("Using object_store credentials configuration");
+            object_store::parse_url_opts(&destination, &snapshots_options.additional_options)?
+                .0
+                .into()
+        };
+
+        // prefix must be stripped of any leading slash and, unless zero-length, end in a single "/" character
+        let prefix: String = destination.path().into();
+        let prefix = match prefix.as_str() {
+            "" | "/" => "".to_string(),
+            prefix => format!("{}/", prefix.trim_start_matches('/').trim_end_matches('/')),
+        };
+
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        snapshot: &PartitionSnapshotMetadata,
+        local_snapshot_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        let partition_id = snapshot.partition_id;
+        let lsn = snapshot.min_applied_lsn;
+
+        debug!(
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        let relative_snapshot_path = format!("lsn_{lsn}", lsn = snapshot.min_applied_lsn);
+        let snapshot_prefix = format!(
+            "{prefix}{partition_id}/{relative_snapshot_path}",
+            prefix = self.prefix,
+        );
+
+        debug!(
+            "Uploading snapshot from {:?} to {}",
+            local_snapshot_path.as_path(),
+            snapshot_prefix
+        );
+
+        for file in &snapshot.files {
+            let filename = file.name.trim_start_matches("/");
+            let key = object_store::path::Path::from(format!(
+                "{}/{}",
+                snapshot_prefix.as_str(),
+                filename
+            ));
+            let put_result = put_snapshot_object(
+                local_snapshot_path.join(filename).as_path(),
+                &key,
+                &self.object_store,
+            )
+            .await?;
+            debug!(
+                etag = put_result.e_tag.unwrap_or_default(),
+                ?key,
+                "Put snapshot data file completed",
+            );
+        }
+
+        let metadata_key =
+            object_store::path::Path::from(format!("{}/metadata.json", snapshot_prefix.as_str()));
+        let metadata_json_payload = PutPayload::from(
+            serde_json::to_string_pretty(snapshot).expect("Can always serialize JSON"),
+        );
+        let put_result = self
+            .object_store
+            .put(&metadata_key, metadata_json_payload)
+            .await?;
```

Comment:
Is there a possibility for two processes taking a snapshot for the same lsn (e.g. an old leader and a new one) which aren't exactly the same because the effective lsn is different? If this is possible, is this a problem?